In [11]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import statsmodels.api as sm

# read data

In [12]:
co=pd.read_csv('owid-co2-data.csv')
temp=pd.read_csv('city_temperature.csv')

/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3185: DtypeWarning: Columns (2) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [13]:
co.head()

iso_code      country  year    co2  co2_per_capita  trade_co2  cement_co2  \
0      AFG  Afghanistan  1949  0.015           0.002        NaN         NaN   
1      AFG  Afghanistan  1950  0.084           0.011        NaN         NaN   
2      AFG  Afghanistan  1951  0.092           0.012        NaN         NaN   
3      AFG  Afghanistan  1952  0.092           0.012        NaN         NaN   
4      AFG  Afghanistan  1953  0.106           0.013        NaN         NaN   

   cement_co2_per_capita  coal_co2  coal_co2_per_capita  ...  \
0                    NaN     0.015                0.002  ...   
1                    NaN     0.021                0.003  ...   
2                    NaN     0.026                0.003  ...   
3                    NaN     0.032                0.004  ...   
4                    NaN     0.038                0.005  ...   

   ghg_excluding_lucf_per_capita  methane  methane_per_capita  nitrous_oxide  \
0                            NaN      NaN                 NaN            NaN   
1                            NaN      NaN                 NaN            NaN   
2                            NaN      NaN                 NaN            NaN   
3                            NaN      NaN                 NaN            NaN   
4                            NaN      NaN                 NaN            NaN   

   nitrous_oxide_per_capita  population           gdp  \
0                       NaN   7624058.0           NaN   
1                       NaN   7752117.0  9.421400e+09   
2                       NaN   7840151.0  9.692280e+09   
3                       NaN   7935996.0  1.001733e+10   
4                       NaN   8039684.0  1.063052e+10   

   primary_energy_consumption  energy_per_capita  energy_per_gdp  
0                         NaN                NaN             NaN  
1                         NaN                NaN             NaN  
2                         NaN                NaN             NaN  
3                         NaN                NaN             NaN  
4                         NaN                NaN             NaN  

[5 rows x 60 columns]

In [14]:
temp.head()

Region  Country State     City  Month  Day  Year  AvgTemperature
0  Africa  Algeria   NaN  Algiers      1    1  1995            64.2
1  Africa  Algeria   NaN  Algiers      1    2  1995            49.4
2  Africa  Algeria   NaN  Algiers      1    3  1995            48.8
3  Africa  Algeria   NaN  Algiers      1    4  1995            46.4
4  Africa  Algeria   NaN  Algiers      1    5  1995            47.9

# merge dataset

In [15]:
temp.columns=temp.columns.map(lambda x: x.lower())

In [16]:
df=pd.merge(co,temp,on=['country','year'])

In [17]:
df.head()

iso_code  country  year    co2  co2_per_capita  trade_co2  cement_co2  \
0      ALB  Albania  1995  2.066           0.664      0.407       0.081   
1      ALB  Albania  1995  2.066           0.664      0.407       0.081   
2      ALB  Albania  1995  2.066           0.664      0.407       0.081   
3      ALB  Albania  1995  2.066           0.664      0.407       0.081   
4      ALB  Albania  1995  2.066           0.664      0.407       0.081   

   cement_co2_per_capita  coal_co2  coal_co2_per_capita  ...           gdp  \
0                  0.026     0.161                0.052  ...  1.160073e+10   
1                  0.026     0.161                0.052  ...  1.160073e+10   
2                  0.026     0.161                0.052  ...  1.160073e+10   
3                  0.026     0.161                0.052  ...  1.160073e+10   
4                  0.026     0.161                0.052  ...  1.160073e+10   

   primary_energy_consumption  energy_per_capita  energy_per_gdp  region  \
0                      22.728           7301.038           1.959  Europe   
1                      22.728           7301.038           1.959  Europe   
2                      22.728           7301.038           1.959  Europe   
3                      22.728           7301.038           1.959  Europe   
4                      22.728           7301.038           1.959  Europe   

   state    city  month  day  avgtemperature  
0    NaN  Tirana      1    1           -99.0  
1    NaN  Tirana      1    2           -99.0  
2    NaN  Tirana      1    3           -99.0  
3    NaN  Tirana      1    4           -99.0  
4    NaN  Tirana      1    5           -99.0  

[5 rows x 66 columns]

# variable correlation

In [23]:
corr=df.corr()
corr

year       co2  co2_per_capita  trade_co2  cement_co2  \
year               1.000000  0.092439       -0.033174  -0.059628    0.103644   
co2                0.092439  1.000000        0.036697  -0.892557    0.984495   
co2_per_capita    -0.033174  0.036697        1.000000   0.037417   -0.033335   
trade_co2         -0.059628 -0.892557        0.037417   1.000000   -0.909441   
cement_co2         0.103644  0.984495       -0.033335  -0.909441    1.000000   
...                     ...       ...             ...        ...         ...   
energy_per_capita -0.007588 -0.021848        0.900855   0.098558   -0.081878   
energy_per_gdp    -0.240850  0.087644        0.468491  -0.168301    0.052564   
month             -0.026811 -0.001859        0.000761  -0.000518   -0.002241   
day               -0.002546 -0.000157        0.000124  -0.000046   -0.000218   
avgtemperature     0.148089  0.019823        0.038342  -0.046379    0.028788   

                   cement_co2_per_capita  coal_co2  coal_co2_per_capita  \
year                            0.056025  0.093235            -0.024886   
co2                             0.259536  0.991489             0.252301   
co2_per_capita                  0.565318 -0.000958             0.700631   
trade_co2                      -0.201502 -0.921662            -0.194653   
cement_co2                      0.267536  0.993608             0.186397   
...                                  ...       ...                  ...   
energy_per_capita               0.530484 -0.080273             0.392714   
energy_per_gdp                  0.145265  0.099037             0.270475   
month                          -0.001338 -0.001520             0.000697   
day                            -0.000107 -0.000135             0.000105   
avgtemperature                  0.087294  0.034107            -0.068861   

                   flaring_co2  flaring_co2_per_capita  ...  nitrous_oxide  \
year                  0.008174               -0.046229  ...       0.047325   
co2                   0.202175               -0.112847  ...       0.907433   
co2_per_capita        0.128303                0.467967  ...      -0.034486   
trade_co2            -0.307747               -0.141926  ...      -0.829281   
cement_co2           -0.034628               -0.172200  ...       0.886733   
...                        ...                     ...  ...            ...   
energy_per_capita     0.125392                0.403801  ...      -0.091098   
energy_per_gdp        0.204336                0.226250  ...       0.031774   
month                -0.002409                0.000403  ...       0.000287   
day                  -0.000180                0.000047  ...       0.000048   
avgtemperature       -0.032774                0.027773  ...       0.057626   

                   nitrous_oxide_per_capita  population       gdp  \
year                              -0.044972    0.043627  0.150191   
co2                               -0.074403    0.814027  0.954167   
co2_per_capita                     0.221396   -0.124561  0.029709   
trade_co2                          0.047885   -0.736964 -0.752028   
cement_co2                        -0.094490    0.785571  0.914076   
...                                     ...         ...       ...   
energy_per_capita                  0.187972   -0.167601 -0.023057   
energy_per_gdp                     0.043905   -0.002200  0.016975   
month                             -0.000074   -0.000704 -0.000092   
day                               -0.000014   -0.000067  0.000043   
avgtemperature                    -0.001798    0.076226  0.031725   

                   primary_energy_consumption  energy_per_capita  \
year                                 0.100346          -0.007588   
co2                                  0.993038          -0.021848   
co2_per_capita                       0.072779           0.900855   
trade_co2                           -0.860528           0.098558   
cement_co2                           0.966718          

# data cleaning

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1374972 entries, 0 to 1374971
Data columns (total 66 columns):
 #   Column                               Non-Null Count    Dtype  
---  ------                               --------------    -----  
 0   iso_code                             1374972 non-null  object 
 1   country                              1374972 non-null  object 
 2   year                                 1374972 non-null  int64  
 3   co2                                  1374972 non-null  float64
 4   co2_per_capita                       1374972 non-null  float64
 5   trade_co2                            1151824 non-null  float64
 6   cement_co2                           1265433 non-null  float64
 7   cement_co2_per_capita                1265433 non-null  float64
 8   coal_co2                             1114413 non-null  float64
 9   coal_co2_per_capita                  1114413 non-null  float64
 10  flaring_co2                          785990 non-null   float64
 11

In [33]:
df.drop_duplicates()

iso_code  country  year    co2  co2_per_capita  trade_co2  cement_co2  \
0            ALB  Albania  1995  2.066           0.664      0.407       0.081   
1            ALB  Albania  1995  2.066           0.664      0.407       0.081   
2            ALB  Albania  1995  2.066           0.664      0.407       0.081   
3            ALB  Albania  1995  2.066           0.664      0.407       0.081   
4            ALB  Albania  1995  2.066           0.664      0.407       0.081   
...          ...      ...   ...    ...             ...        ...         ...   
1374967      ZMB   Zambia  2014  4.877           0.317      4.113       0.736   
1374968      ZMB   Zambia  2014  4.877           0.317      4.113       0.736   
1374969      ZMB   Zambia  2014  4.877           0.317      4.113       0.736   
1374970      ZMB   Zambia  2014  4.877           0.317      4.113       0.736   
1374971      ZMB   Zambia  2014  4.877           0.317      4.113       0.736   

         cement_co2_per_capita  coal_co2  coal_co2_per_capita  ...  \
0                        0.026     0.161                0.052  ...   
1                        0.026     0.161                0.052  ...   
2                        0.026     0.161                0.052  ...   
3                        0.026     0.161                0.052  ...   
4                        0.026     0.161                0.052  ...   
...                        ...       ...                  ...  ...   
1374967                  0.048     0.484                0.031  ...   
1374968                  0.048     0.484                0.031  ...   
1374969                  0.048     0.484                0.031  ...   
1374970                  0.048     0.484                0.031  ...   
1374971                  0.048     0.484                0.031  ...   

                  gdp  primary_energy_consumption  energy_per_capita  \
0        1.160073e+10                      22.728           7301.038   
1        1.160073e+10                      22.728           7301.038   
2        1.160073e+10                      22.728           7301.038   
3        1.160073e+10                      22.728           7301.038   
4        1.160073e+10                      22.728           7301.038   
...               ...                         ...                ...   
1374967  5.341918e+10                      52.303           3396.329   
1374968  5.341918e+10                      52.303           3396.329   
1374969  5.341918e+10                      52.303           3396.329   
1374970  5.341918e+10                      52.303           3396.329   
1374971  5.341918e+10                      52.303           3396.329   

         energy_per_gdp  region  state    city  month  day  avgtemperature  
0                 1.959  Europe    NaN  Tirana      1    1           -99.0  
1                 1.959  Europe    NaN  Tirana      1    2           -99.0  
2                 1.959  Europe    NaN  Tirana      1    3           -99.0  
3                 1.959  Europe    NaN  Tirana      1    4           -99.0  
4                 1.959  Europe    NaN  Tirana      1    5           -99.0  
...                 ...     ...    ...     ...    ...  ...             ...  
1374967           0.979  Africa    NaN  Lusaka      1   18            71.3  
1374968           0.979  Africa    NaN  Lusaka      1   19            77.5  
1374969           0.979  Africa    NaN  Lusaka      1   20            73.4  
1374970           0.979  Africa    NaN  Lusaka      1   21            73.1  
1374971           0.979  Africa    NaN  Lusaka      1   22            71.3  

[1372932 rows x 66 columns]

# model 

### 1. Simple linear regression explores the relationship between average temperature and co2

In [18]:
fit1=sm.formula.ols('avgtemperature~co2',data=df).fit()
print(fit1.params)

Intercept    55.524510
co2           0.000596
dtype: float64


In [21]:
print(fit1.summary())

                            OLS Regression Results                            
Dep. Variable:         avgtemperature   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     540.5
Date:                Sun, 13 Nov 2022   Prob (F-statistic):          1.53e-119
Time:                        17:05:30   Log-Likelihood:            -7.0454e+06
No. Observations:             1374972   AIC:                         1.409e+07
Df Residuals:                 1374970   BIC:                         1.409e+07
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     55.5245      0.037   1514.538      0.0

From the summary above, it can be observed that p-value is 0.000, which means that Co2 emissions have significant effect on average temperature. Moreover, the slop of Co2 is 0.000596, which means that the relationship between Co2 emissions and average temperature is positive.  

### 2. Multiple linear regression to explore whether different areas of the planet are affected differently by emission increase

In [25]:
fit2=sm.formula.ols('avgtemperature~co2+country+co2*country',data=df).fit()
print(fit2.params)

Intercept                   -127.571154
country[T.Algeria]           192.573417
country[T.Argentina]         186.901289
country[T.Australia]         176.588410
country[T.Austria]           177.191710
                                ...    
co2:country[T.Uruguay]       -40.605397
co2:country[T.Uzbekistan]    -41.077033
co2:country[T.Venezuela]     -41.035198
co2:country[T.Vietnam]       -41.079637
co2:country[T.Zambia]        -41.778164
Length: 232, dtype: float64


In [27]:
print(fit2.summary())

                            OLS Regression Results                            
Dep. Variable:         avgtemperature   R-squared:                       0.303
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                     2590.
Date:                Sun, 13 Nov 2022   Prob (F-statistic):               0.00
Time:                        17:27:49   Log-Likelihood:            -6.7973e+06
No. Observations:             1374972   AIC:                         1.360e+07
Df Residuals:                 1374740   BIC:                         1.360e+07
Df Model:                         231                                         
Covariance Type:            nonrobust                                         
                                              coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------

In this model, we use Co2 emissions, countries, and the intersection of Co2 emissions and countries as independent variables. In the summary table, all the p-value of variables are 0.000, which means that Co2 emissions, countries, and the intersection of Co2 emissions and countries all have significant effect on average temperature. Thus, different areas of the planet are affected differently by emission increase